src/ppx/application.cpp

```diff
@@ -1491,6 +1415,142 @@ int Application::Run(int argc, char** argv)
             Quit();
         }
     }
+}
+
+int Application::Run(int argc, char** argv)
+{
+    // Only allow one instance of Application. Since we can't stop
+    // the app in the ctor - stop it here.
+    //
+    if (this != sApplicationInstance) {
+        return false;
+    }
+
+    // Parse args.
+    if (Failed(mCommandLineParser.Parse(argc, const_cast<const char**>(argv)))) {
+        PPX_ASSERT_MSG(false, "Unable to parse command line arguments");
+        return EXIT_FAILURE;
+    }
+
+    // Call config.
+    // Put this early because it might disable the display.
+    Config(mSettings);
+
+    // Knobs need to be set up after commandline parsing.
+    // This has dependency on Config() where standard knob default values are set
```

Comment:
There seems to be a lot of dependencies with config, knobs and command line parser in this section which seems a bit brittle. In particular here, we have to call Config() since it sets defaults, then initialize knobs and then finalize settings since they are updated from the knobs. Have we considered settings defaults for knobs in the InitKnobs instead of in Config()? https://github.com/google/bigwheels/pull/362/files
Just a thought though, thinking of how to make this code clearer

Reply:
Agree, it is really error prone like this. 
Thanks for sharing the link of the PR, looks like a better solution. Although I feel that the name `InitKnob` is not quite right: the function is allowing the derived class to have its own knob initializations, plus changing the default values of the standard knobs. Would it be clearer to have another separate virtual function purely for allowing the derived class to overwrite the default value of the standard knobs? Not sure if this is better, would like to hear your opinion on this. thanks :)

Reply:
Hm, I don't think I have a strong opinion on that one. I can see an argument for either way, so I will be fine with either. I think there is other general weirdness with mSettings and knobs, but we can address that separately including the default options in Config() or somewhere else. Just want to explore that before its use solidified.